# Keras starter kit [full training set, UNet]

In [407]:
# import sys

# sys.path.append('/kaggle/input/pretrainedmodels/pretrainedmodels-0.7.4')
# sys.path.append('/kaggle/input/segmentation-models-pytorch/segmentation_models.pytorch-master')
# sys.path.append('/kaggle/input/efficientnet-pytorch/EfficientNet-PyTorch-master')
# sys.path.append('/kaggle/input/timm-pytorch-image-models/pytorch-image-models-master')
# sys.path.append("/kaggle/input/einops/einops-master")

In [408]:
!export CUDA_VISIBLE_DEVICES=0

/bin/bash: /home/fummicc1/anaconda3/lib/libtinfo.so.6: no version information available (required by /bin/bash)


## Setup

In [409]:
import numpy as np
import torch
import torch.nn as nn
import wandb
import torchvision
import datetime
import imageio

# import cupy
from sklearn.model_selection import KFold
import albumentations as A
from albumentations.pytorch import ToTensorV2
import pytorch_lightning
import segmentation_models_pytorch as smp
import pytorch_lightning as pl
import pytorch_lightning.callbacks.model_checkpoint
import pytorch_lightning.plugins
from skimage.transform import resize as resize_ski
from pytorch_lightning.strategies.ddp import DDPStrategy
from pytorch_lightning.loggers import WandbLogger
from einops import rearrange, reduce, repeat
import torch.nn.functional as F
import segmentation_models_pytorch as smp
from segmentation_models_pytorch.decoders.unet.decoder import UnetDecoder, DecoderBlock
from timm.models.resnet import (
    resnet10t,
    resnet34d,
    resnet50d,
    resnet14t,
    seresnext26d_32x4d,
    seresnext50_32x4d,
    seresnext26tn_32x4d,
)
from timm.models.mvitv2 import mvitv2_base
import os
import torch.utils.data
from dataclasses import dataclass

from scipy.ndimage import distance_transform_edt

import ssl

ssl._create_default_https_context = ssl._create_unverified_context

import glob
import matplotlib.pyplot as plt
from tqdm import tqdm
import cv2
import torch
import torch.nn as nn
import torch.optim as optim
import torch.utils.data
import os, cv2
import gc
import sys
import matplotlib.patches as patches
import random
import numpy as np
import pandas as pd
from tqdm import tqdm
import matplotlib.pyplot as plt
import pytorch_lightning as pl

import torch
import torch.optim as optim
import torch.nn as nn
import torch.nn.functional as F
from torch.cuda import amp
from torch.utils.data import Dataset, DataLoader
import segmentation_models_pytorch as smp
import albumentations as A
from IPython.display import Video

pytorch_lightning.seed_everything(seed=42)
torch.set_float32_matmul_precision("high")


@dataclass
class CFG:
    # Data config
    # DATA_DIR = '/kaggle/input/vesuvius-challenge-ink-detection/'
    DATA_DIR = "/home/fummicc1/codes/competitions/kaggle-ink-detection"
    # DATA_DIR = '/home/fummicc1/codes/Kaggle/kaggle-ink-detection'
    BUFFER = 192  # Half-size of papyrus patches we'll use as model inputs
    CROP_SIZE = BUFFER * 2 # 384px
    STRIDE = 128 # 128
    Z_LIST = list(range(20, 40, 2)) # 10チャネル
    # Z_LIST = list(range(4, 16, 2)) + list(range(16, 36)) + list(range(36, 44, 2))
    Z_DIM = len(
        Z_LIST
    )  # Number of slices in the z direction. Max value is 64 - Z_START
    BATCH_Z_DIFF = 4
    BATCH_Z_INFO = [
        # (start, end)
        (0, 4),
        (2, 6),
        (4, 8),
        (6, 10),
    ]
    SHARED_HEIGHT = None  # Max height to resize all papyrii

    # Model config
    BATCH_SIZE = 1

    GPU_COUNT = 4

    device = torch.device("cuda")
    threshold = 0.5
    num_workers = 0
    exp = 1e-7
    mask_padding = (SHARED_HEIGHT // 80) if SHARED_HEIGHT is not None else 100

    num_epochs = 40
    lr = 1e-3 # initial lr
    lr_scheduler_name = "CosineAnnealingLR"
    eta_min_lr = 1e-5
    WANDB_NOTE = "data_listの順番を変更して[1, 3, 2a, 2b]になった."
    augmentation_names = [
        "Horizontal",
        "Vertical",
        "RandomScale",
        "Transpose",
        "RandomRotate",
        "ShiftScaleRotate",
        # "Blur",
        "GridDistortion",
        "MultiChannelNoise",
        "PadIfNeeded",
        "Resize",
        "MixedUp",
    ]
    
    mixed_up_alpha = 0.9
    mixed_up_probability = 0.4
    mixed_up_epochs = list(range(0, 7)) + list(range(num_epochs - 12, num_epochs - 7))
    
    loss1_alpha = 0.8
    loss1_beta = None

    loss2a_alpha = 1.2
    loss2a_beta = None

    loss2b_alpha = 1.2

    loss1_weight = 0.5
    loss2_weight = 0.5

    loss2a_weight = 0.5
    loss2b_weight = 0.5

    lr_scheduler = optim.lr_scheduler.CosineAnnealingLR
    optimizer = "AdamW"
    # loss1 = smp.losses.TverskyLoss(
    #     smp.losses.BINARY_MODE,
    #     log_loss=False,
    #     from_logits=True,
    #     smooth=1e-7,
    #     alpha=loss1_alpha,
    #     beta=loss1_beta,
    # )
    loss1 = smp.losses.SoftBCEWithLogitsLoss(
        pos_weight=torch.tensor([loss1_alpha]),
    )
    # loss2a = smp.losses.TverskyLoss(
    #     smp.losses.BINARY_MODE,
    #     log_loss=False,
    #     from_logits=True,
    #     smooth=1e-7,
    #     alpha=loss2a_alpha,
    #     beta=loss2a_beta,
    # )
    loss2a = smp.losses.SoftBCEWithLogitsLoss(
        pos_weight=torch.tensor([loss2a_alpha]),
    )
    loss2b = nn.BCEWithLogitsLoss(
        pos_weight=torch.tensor([loss2b_alpha]),
    )

    noise_intensity = (0.000025, 0.00005)
    
    mean = 0.42
    std = 0.3
    
    use_new_label_mask = False
    pretrained = False
    init_weight_bias = False
    SKIP_ZERO_ON_TRAIN = False
    model_name = "seresnext26d_32x4d"
    # MODEL_DEPTH = 34
    # MODEL_KIND = "KM"
    
    precision = "float32" # float16

    interpolation = "bicubic" # PixelShuffling
    # use_pixel_shuffle = True

    exp_name = "123-gpu17"
    prev_exp_name = "102-gpu17"
    
    attention_type = None


def class2dict(c):
    return {
        attr: getattr(c, attr)
        for attr in dir(c)
        if not callable(getattr(c, attr)) and not attr.startswith("__")
    }

Global seed set to 42


In [410]:
import cupy as cp
xp = cp

delta_lookup = {
    "xx": xp.array([[1, -2, 1]], dtype=float),
    "yy": xp.array([[1], [-2], [1]], dtype=float),
    "xy": xp.array([[1, -1], [-1, 1]], dtype=float),
}

def operate_derivative(img_shape, pair):
    assert len(img_shape) == 2
    delta = delta_lookup[pair]
    fft = xp.fft.fftn(delta, img_shape)
    return fft * xp.conj(fft)

def soft_threshold(vector, threshold):
    return xp.sign(vector) * xp.maximum(xp.abs(vector) - threshold, 0)

def back_diff(input_image, dim):
    assert dim in (0, 1)
    r, n = xp.shape(input_image)
    size = xp.array((r, n))
    position = xp.zeros(2, dtype=int)
    temp1 = xp.zeros((r+1, n+1), dtype=float)
    temp2 = xp.zeros((r+1, n+1), dtype=float)
    
    temp1[position[0]:size[0], position[1]:size[1]] = input_image
    temp2[position[0]:size[0], position[1]:size[1]] = input_image
    
    size[dim] += 1
    position[dim] += 1
    temp2[position[0]:size[0], position[1]:size[1]] = input_image
    temp1 -= temp2
    size[dim] -= 1
    return temp1[0:size[0], 0:size[1]]


def forward_diff(input_image, dim):
    assert dim in (0, 1)
    r, n = xp.shape(input_image)
    size = xp.array((r, n))
    position = xp.zeros(2, dtype=int)
    temp1 = xp.zeros((r+1, n+1), dtype=float)
    temp2 = xp.zeros((r+1, n+1), dtype=float)
        
    size[dim] += 1
    position[dim] += 1

    temp1[position[0]:size[0], position[1]:size[1]] = input_image
    temp2[position[0]:size[0], position[1]:size[1]] = input_image
    
    size[dim] -= 1
    temp2[0:size[0], 0:size[1]] = input_image
    temp1 -= temp2
    size[dim] += 1
    return -temp1[position[0]:size[0], position[1]:size[1]]

def iter_deriv(input_image, b, scale, mu, dim1, dim2):
    g = back_diff(forward_diff(input_image, dim1), dim2)
    d = soft_threshold(g + b, 1 / mu)
    b = b + (g - d)
    L = scale * back_diff(forward_diff(d - b, dim2), dim1)
    return L, b

def iter_xx(*args):
    return iter_deriv(*args, dim1=1, dim2=1)

def iter_yy(*args):
    return iter_deriv(*args, dim1=0, dim2=0)

def iter_xy(*args):
    return iter_deriv(*args, dim1=0, dim2=1)

def iter_sparse(input_image, bsparse, scale, mu):
    d = soft_threshold(input_image + bsparse, 1 / mu)
    bsparse = bsparse + (input_image - d)
    Lsparse = scale * (d - bsparse)
    return Lsparse, bsparse

def denoise_image(input_image, iter_num=100, fidelity=150, sparsity_scale=10, continuity_scale=0.5, mu=1):
    image_size = xp.shape(input_image)
    #print("Initialize denoising")
    norm_array = (
        operate_derivative(image_size, "xx") + 
        operate_derivative(image_size, "yy") + 
        2 * operate_derivative(image_size, "xy")
    )
    norm_array += (fidelity / mu) + sparsity_scale ** 2
    b_arrays = {
        "xx": xp.zeros(image_size, dtype=float),
        "yy": xp.zeros(image_size, dtype=float),
        "xy": xp.zeros(image_size, dtype=float),
        "L1": xp.zeros(image_size, dtype=float),
    }
    g_update = xp.multiply(fidelity / mu, input_image)
    for i in tqdm(range(iter_num), total=iter_num):
        #print(f"Starting iteration {i+1}")
        g_update = xp.fft.fftn(g_update)
        if i == 0:
            g = xp.fft.ifftn(g_update / (fidelity / mu)).real
        else:
            g = xp.fft.ifftn(xp.divide(g_update, norm_array)).real
        g_update = xp.multiply((fidelity / mu), input_image)
        
        #print("XX update")
        L, b_arrays["xx"] = iter_xx(g, b_arrays["xx"], continuity_scale, mu)
        g_update += L
        
        #print("YY update")
        L, b_arrays["yy"] = iter_yy(g, b_arrays["yy"], continuity_scale, mu)
        g_update += L
        
        #print("XY update")
        L, b_arrays["xy"] = iter_xy(g, b_arrays["xy"], 2 * continuity_scale, mu)
        g_update += L
        
        #print("L1 update")
        L, b_arrays["L1"] = iter_sparse(g, b_arrays["L1"], sparsity_scale, mu)
        g_update += L
        
    g_update = xp.fft.fftn(g_update)
    g = xp.fft.ifftn(xp.divide(g_update, norm_array)).real
    
    g[g < 0] = 0
    g -= g.min()
    g /= g.max()
    return g

In [411]:
def is_in_masked_zone(location, mask):
    return mask[location[0], location[1]] > 0


In [412]:
def resize(img):
    current_height, current_width = img.shape    
    aspect_ratio = current_width / current_height
    if CFG.SHARED_HEIGHT is None:
        return img
    # new_height = CFG.SHARED_HEIGHT
    # pad_y = new_height - current_height
    # if pad_y > 0:
    #     # 元画像が小さい場合は解像度を大きくしないでpaddingをつける
    #     img = np.pad(img, [(0, pad_y), (0, 0)], constant_values=0)
    # else:
    # 既に十分でかい場合はリサイズする
    # 本当はpaddingしたいけど、メモリサイズが大きくなる
    if CFG.SHARED_HEIGHT > current_height:
        # 既に小さい場合はリサイズしない
        return img    
    new_height = CFG.SHARED_HEIGHT
    new_width = int(CFG.SHARED_HEIGHT * aspect_ratio)
    new_size = (new_width, new_height)
    # (W, H)の順で渡すが結果は(H, W)になっている
    img = cv2.resize(img, new_size)
    return img

def load_mask(split, index):     
    img = cv2.imread(f"{CFG.DATA_DIR}/{split}/{index}/mask.png", 0) // 255
    img = resize(img)
    return img


def load_labels(split, index):    
    suffix = "_new" if CFG.use_new_label_mask else ""
    img = cv2.imread(f"{CFG.DATA_DIR}/{split}/{index}/inklabels{suffix}.png", 0) // 255    
    img = resize(img)
    return img

In [413]:
def extract_subvolume(location, volume):
    global printed
    y = location[0]
    x = location[1]
    subvolume = volume[y-CFG.BUFFER:y+CFG.BUFFER, x-CFG.BUFFER:x+CFG.BUFFER, :].astype(np.float32)
    
    return subvolume

In [414]:
def load_volume(split, index):    
    # Load the 3d x-ray scan, one slice at a time
    all = sorted(glob.glob(f"{CFG.DATA_DIR}/{split}/{index}/surface_volume/*.tif"))
    z_slices_fnames = [all[i] for i in range(len(all)) if i in CFG.Z_LIST]
    assert len(z_slices_fnames) == CFG.Z_DIM
    z_slices = []
    for z, filename in  tqdm(enumerate(z_slices_fnames)):
        img = cv2.imread(filename, -1)        
        img = resize(img)
        # img = (img / (2 ** 8)).astype(np.uint8)
        img = img.astype(np.float32) / 255
        z_slices.append(img)
    return np.stack(z_slices, axis=-1)

In [415]:
def generate_locations_ds(volume, mask, label=None, skip_zero=False):
    is_in_mask_train = lambda x: is_in_masked_zone(x, mask)

    # Create a list to store train locations
    locations = []

    # Generate train locations
    volume_height, volume_width = volume.shape[:-1]

    for y in range(CFG.BUFFER, volume_height - CFG.BUFFER, CFG.STRIDE):
        for x in range(CFG.BUFFER, volume_width - CFG.BUFFER, CFG.STRIDE):
            if skip_zero and CFG.SKIP_ZERO_ON_TRAIN and label is not None and np.all(label[y - CFG.BUFFER : y + CFG.BUFFER, x - CFG.BUFFER : x + CFG.BUFFER] == 0):
                # print(f"skip location at (y: {y}, x: {x})")
                continue
            if is_in_mask_train((y, x)):
                locations.append((y, x))

    # Convert the list of train locations to a PyTorch tensor
    locations_ds = np.stack(locations, axis=0)
    return locations_ds

In [416]:
import torch
import numpy as np

def mixup_data(x, y):
    alpha = CFG.mixed_up_alpha
    if alpha > 0.:
        lam = np.random.beta(alpha, alpha)
    else:
        lam = 1.

    batch_size = x.size()[0]
    index = torch.randperm(batch_size)

    mixed_x = lam * x + (1 - lam) * x[index]    
    y_a, y_b = y, y[index]
    
    mixed_x = mixed_x.float()
    y_a = y_a.float()
    y_b = y_b.float()
    
    y = y_a * lam + y_b * (1 - lam)

    return mixed_x, y, y_a, y_b, lam


In [417]:
import numpy as np
from albumentations.core.transforms_interface import ImageOnlyTransform

class MultichannelNoise(ImageOnlyTransform):

    def __init__(self, intensity=CFG.noise_intensity, always_apply=False, p=0.5):
        super().__init__(always_apply, p)
        self.intensity = intensity

    def apply(self, img, **params):
        intensity = np.random.uniform(*self.intensity)
        noise = np.random.normal(loc=0, scale=intensity, size=img.shape)
        # print("img", img)
        img = img + noise        
        return np.clip(img, 0, 1).astype(np.float32) # クリップして0から255の範囲に保つ


In [418]:

import torch
from torch.utils.data import Dataset, DataLoader
from sklearn.preprocessing import OneHotEncoder

from albumentations.core.transforms_interface import ImageOnlyTransform

class SubvolumeDataset(Dataset):
    def __init__(self, locations, volume, labels, buffer, is_train: bool, return_location: bool = False):
        self.locations = locations
        self.volume = volume
        self.labels = labels        
        self.buffer = buffer
        self.is_train = is_train
        self.return_location = return_location

    def __len__(self):
        return len(self.locations)

    def __getitem__(self, idx):
        label = None
        location = np.array(self.locations[idx])
        y, x = location[0], location[1]

        subvolume = extract_subvolume(location, self.volume)
        
        if self.labels is not None:
            label = self.labels[y - self.buffer:y + self.buffer, x - self.buffer:x + self.buffer]            
            label = np.stack([label], axis=-1)            
            
        # 段々meanは小さくなる
        mean = np.array([CFG.mean for i in range(0, CFG.Z_DIM)]).reshape(-1, 1, 1)
        # 段々stdは小さくなる
        std = np.array([CFG.std for i in range(0, CFG.Z_DIM)]).reshape(-1, 1, 1)
        
        if self.is_train and label is not None:    
            # label = smooth_labels(label, alpha=0.1)                  
            transformed = A.Compose([
                A.ToFloat(max_value=255),
                A.HorizontalFlip(p=0.5) if "Horizontal" in CFG.augmentation_names else A.Compose([]),
                A.VerticalFlip(p=0.5) if "Vertical" in CFG.augmentation_names else A.Compose([]),  
                A.RandomScale(p=0.4, scale_limit=0.3) if "RandomScale" in CFG.augmentation_names else A.Compose([]),
                A.Transpose(p=0.5) if "Transpose" in CFG.augmentation_names else A.Compose([]), 
                A.RandomRotate90(p=0.5,) if "RandomRotate" in CFG.augmentation_names else A.Compose([]), 
                A.ShiftScaleRotate(p=0.7, scale_limit=0.35, rotate_limit=45) if "ShiftScaleRotate" in CFG.augmentation_names else A.Compose([]),
                A.MotionBlur(p=0.2) if "Blur" in CFG.augmentation_names else A.Compose([]),
                A.GridDistortion(num_steps=5, distort_limit=0.15, p=0.3) if "GridDistortion" in CFG.augmentation_names else A.Compose([]),
                MultichannelNoise(
                    p=0.2,
                ) if "MultichannelNoise" in CFG.augmentation_names else A.Compose([]),
                A.PadIfNeeded(min_height=self.buffer * 2, min_width=self.buffer * 2) if "PadIfNeeded" in CFG.augmentation_names else A.Compose([]),
                A.Resize(height=self.buffer * 2, width=self.buffer * 2) if "Resize" in CFG.augmentation_names else A.Compose([]),
            ])(image=subvolume, mask=label)            
            subvolume = transformed["image"]
            label = transformed["mask"]
            subvolume = np.transpose(subvolume, (2, 0, 1))
            label = np.transpose(label, (2, 0, 1))
            subvolume = (subvolume - mean) / std            
        else:
            if label is None:
                subvolume = np.transpose(subvolume, (2, 0, 1))
                subvolume /= 255.
                subvolume = (subvolume - mean) / std
            else:
                # print("subvolume in val dataset (before aug)", subvolume, file=open("before-val-aug.log", "w")) 
                subvolume = np.transpose(subvolume, (2, 0, 1))
                label = np.transpose(label, (2, 0, 1))
                subvolume /= 255.
                subvolume = (subvolume - mean) / std
        # print("subvolume", subvolume)
        if self.return_location:
            return subvolume, location
        return subvolume, label        

In [419]:
def initialize_weights(model: nn.Module):
    for module in model.modules():
        if isinstance(module, (nn.Conv2d, nn.Linear)):
            nn.init.kaiming_normal_(module.weight, nonlinearity='relu')
            if module.bias is not None:
                nn.init.constant_(module.bias, 0)
        elif isinstance(module, nn.BatchNorm2d):
            nn.init.constant_(module.weight, 1)
            nn.init.constant_(module.bias, 0)

In [420]:
def init_subpixel(weight):
    co, ci, h, w = weight.shape
    co2 = co // 4
    # initialize sub kernel
    k = torch.empty([co2, ci, h, w])
    nn.init.kaiming_uniform_(k)
    # repeat 4 times
    k = k.repeat_interleave(4, dim=0)
    weight.data.copy_(k)

In [421]:
def init_linear(m, relu=True):
    if relu: nn.init.kaiming_uniform_(m.weight, mode='fan_in', nonlinearity='relu')
    else: nn.init.xavier_uniform_(m.weight)
    if m.bias is not None: nn.init.zeros_(m.bias)

In [422]:
def initialize_weights(model: nn.Module):
    for module in model.modules():
        if isinstance(module, (nn.Conv2d, nn.Linear)):
            nn.init.kaiming_normal_(module.weight, nonlinearity='relu')
            if module.bias is not None:
                nn.init.constant_(module.bias, 0)
        elif isinstance(module, nn.BatchNorm2d):
            nn.init.constant_(module.weight, 1)
            nn.init.constant_(module.bias, 0)

In [423]:
class SmpUnetDecoder(nn.Module):
	def __init__(
		self,
		in_channel,
		skip_channel,
		out_channel,
	):
		super().__init__()
		self.center = nn.Identity()

		i_channel = [
			in_channel,
		] + out_channel[:-1]
		s_channel = skip_channel
		o_channel = out_channel
		# print("i:", i_channel)
		# print("s:", s_channel)
		# print("o:", o_channel)
		block = [
			DecoderBlock(i, s, o, use_batchnorm=True, attention_type=CFG.attention_type)
			for i, s, o in zip(i_channel, s_channel, o_channel)
		]
		self.block = nn.ModuleList(block)

	def forward(self, feature, skip):
		# for s in skip:
			# print("feature", feature.shape, "skip", s.shape)
		d = self.center(feature)
		decode = []
		for i, block in enumerate(self.block):
			s = skip[i]
			d = block(d, s)
			decode.append(d)

		last = d
		return last, decode

class Net(nn.Module):
	def __init__(
		self,
	):
		super().__init__()
		self.output_type = ["inference", "loss"]

		conv_dim = 64
		encoder1_dim = [
			conv_dim,
			256,
			512,
			1024,
			2048,
		]
		decoder1_dim = [
			1024,
			512,
			256,
			64,
		]
		if CFG.model_name == "seresnext50_32x4d":
			self.encoder1 = seresnext50_32x4d(pretrained=CFG.pretrained, in_chans=CFG.BATCH_Z_DIFF)
		elif CFG.model_name == "seresnext26d_32x4d":
			self.encoder1 = seresnext26d_32x4d(pretrained=CFG.pretrained, in_chans=CFG.BATCH_Z_DIFF)
		elif CFG.model_name == "seresnext26tn_32x4d":
			self.encoder1 = seresnext26tn_32x4d(pretrained=CFG.pretrained, in_chans=CFG.BATCH_Z_DIFF)

		self.decoder1 = SmpUnetDecoder(
			in_channel=encoder1_dim[-1],
			skip_channel=encoder1_dim[:-1][::-1],
			out_channel=decoder1_dim,
		)
		# -- pool attention weight  
		self.weight1 = nn.ModuleList(
			[
				nn.Sequential(
					nn.Conv2d(dim, dim, kernel_size=3, padding=1),
					nn.BatchNorm2d(dim),
					nn.ReLU(inplace=True),
				)
				for dim in encoder1_dim
			]
		)
		self.logit1 = nn.Conv2d(decoder1_dim[-1], 1, kernel_size=1)

		# --------------------------------		
		encoder2_dim = [64, 64, 128, 256, 512]
		decoder2_dim = [
			256,
			128,
			64,
			64,
		]
		self.encoder2 = resnet10t(pretrained=CFG.pretrained, in_chans=decoder1_dim[-1])

		self.decoder2 = SmpUnetDecoder(
			in_channel=encoder2_dim[-1],
			skip_channel=encoder2_dim[:-1][::-1],
			out_channel=decoder2_dim,
		)
		self.logit2 = nn.Conv2d(decoder2_dim[-1], 1, kernel_size=1)

		if CFG.init_weight_bias:
			initialize_weights(self.logit1)
			initialize_weights(self.logit2)
			initialize_weights(self.weight1)

	def forward(self, batch):
		v = batch
		B, C, H, W = v.shape
		vv = [
			v[:, a: b]
			for a, b in CFG.BATCH_Z_INFO
		]
		# print(list(map(lambda a: a.shape, vv)))
		K = len(vv)
		x = torch.cat(vv, 0)
		# x = v

		# ----------------------
		encoder = []
		e = self.encoder1
		x = e.conv1(x)
		x = e.bn1(x)
		x = e.act1(x)
		encoder.append(x)
		x = F.avg_pool2d(x, kernel_size=2, stride=2)
		x = e.layer1(x)
		encoder.append(x)
		x = e.layer2(x)
		encoder.append(x)
		x = e.layer3(x)
		encoder.append(x)
		x = e.layer4(x)
		encoder.append(x)
		# print('encoder', [f.shape for f in encoder])

		for i in range(len(encoder)):
			e = encoder[i]
			e = F.avg_pool2d(e, 2, 2)
			f = self.weight1[i](e)
			_, c, h, w = e.shape
			f = rearrange(f, "(K B) c h w -> B K c h w", K=K, B=B, h=h, w=w)  #
			e = rearrange(e, "(K B) c h w -> B K c h w", K=K, B=B, h=h, w=w)  #
			w = F.softmax(f, 1)
			e = (w * e).sum(1)
			encoder[i] = e

		feature = encoder[-1]
		skip = encoder[:-1][::-1]
		# print('encoder after weighted', [f.shape for f in encoder])
		# print('skip after weighted', [f.shape for f in skip])
		last, decoder = self.decoder1(feature, skip)
		logit1 = self.logit1(last)
		# print("last.shape", last.shape)
		# print("logit1.shape", logit1.shape)
		# print("encoder1 shapes", list(map(lambda a: a.shape, encoder)))
  
		logit1 = F.interpolate(
			logit1, size=(H, W), mode=CFG.interpolation, align_corners=False, antialias=True
		)

		# ----------------------
		x = last  # .detach()
		# x = F.avg_pool2d(x,kernel_size=2,stride=2)
		encoder = []
		e = self.encoder2
		x = e.conv1(x)
		x = e.bn1(x)
		x = e.act1(x)
		encoder.append(x)
		x = F.avg_pool2d(x, kernel_size=2, stride=2)
		x = e.layer1(x)
		encoder.append(x)
		x = e.layer2(x)
		encoder.append(x)
		x = e.layer3(x)
		encoder.append(x)
		x = e.layer4(x)
		encoder.append(x)
		
		feature = encoder[-1]
		skip = encoder[:-1][::-1]

		# print("feature.shape", list(map(lambda a: a.shape, feature)))
		# print("skip.shape", list(map(lambda a: a.shape, skip)))

		last, decoder = self.decoder2(feature, skip)
		logit2 = self.logit2(last)
		# print("logit2.shape", logit2.shape)
		logit2 = F.interpolate(
			logit2, size=(H, W), mode=CFG.interpolation, align_corners=False, antialias=True
		)
		return logit1, logit2
		# return logit1

In [424]:
# ref - https://www.kaggle.com/competitions/vesuvius-challenge-ink-detection/discussion/397288
def fbeta_score(preds, targets, threshold, beta=0.5, smooth=1e-5):
    preds_t = torch.where(preds > threshold, 1.0, 0.0).float()
    y_true_count = targets.sum()
    
    ctp = preds_t[targets==1].sum()
    cfp = preds_t[targets==0].sum()
    beta_squared = beta * beta

    c_precision = ctp / (ctp + cfp + smooth)
    c_recall = ctp / (y_true_count + smooth)
    dice = (1 + beta_squared) * (c_precision * c_recall) / (beta_squared * c_precision + c_recall + smooth)

    return dice

In [425]:
tc = torch
def TTA(x:tc.Tensor,model:nn.Module):
    #x.shape=(batch,c,h,w)
    shape=x.shape
    x=[x,*[tc.rot90(x,k=i,dims=(-2,-1)) for i in range(1,4)]]
    x=tc.cat(x,dim=0)
    _, x = model(x)
    x=x.reshape(4,shape[0], 1 ,*shape[2:])
    x=[tc.rot90(x[i],k=-i,dims=(-2,-1)) for i in range(4)]
    x=tc.stack(x,dim=0)
    return x.mean(0)

In [426]:


class Model(pl.LightningModule):
	training_step_outputs = []
	validation_step_outputs = []
	test_step_outputs = [[], []]

	def __init__(self, **kwargs):
		super().__init__()

		self.model = Net()        

		self.loss1 = CFG.loss1
		self.loss2 = CFG.loss2a
		self.loss3 = CFG.loss2b

	def forward(self, image, stage):
		if stage != "train":
			mask = TTA(image, self.model)
			# _, mask = self.model(image)
		else:
			mask = self.model(image)
		return mask

	def shared_step(self, batch, stage):
		subvolumes, labels = batch

		image, labels = subvolumes.float(), labels.float()        
		assert image.ndim == 4
		
		h, w = image.shape[2:]
		assert h % 32 == 0 and w % 32 == 0
		
		# print("labels", labels.max(), labels.min())

		assert labels.max() <= 1.0 and labels.min() >= 0

		if stage == "train":
			
			# decide if perform mixed-up augmentation
			should_mixed_up = "MixedUp" in CFG.augmentation_names and  np.random.rand(1)[0] <= CFG.mixed_up_probability and self.current_epoch in range(CFG.start_mixedup_epochs, CFG.stop_mixedup_epochs)
			if should_mixed_up:
				# perform mixed-up
				image, labels, labels_a, labels_b, lam = mixup_data(subvolumes, labels)
			
			
			logit1, logit2 = self.forward(image, stage)
			if should_mixed_up:
				loss1 = self.loss1(logit1, labels)
				loss2a = self.loss2(logit2, labels)
				loss2b = self.loss3(logit2, labels_a) * lam + self.loss3(logit2, labels_b) * (1 - lam)
				# loss2b = self.loss3(logit2, labels)
				loss = CFG.loss1_weight * loss1 + CFG.loss2_weight * (loss2a * CFG.loss2a_weight + loss2b * CFG.loss2b_weight)
			else:
				loss = self.loss1(logit1, labels) * CFG.loss1_weight + CFG.loss2_weight * (self.loss2(logit2, labels) * CFG.loss2a_weight + self.loss3(logit2, labels) * CFG.loss2b_weight)
			# loss = self.loss2(logit, labels) * CFG.loss2a_weight + self.loss3(logit, labels) * CFG.loss2b_weight
			logit = logit2
		else:
			logit = self.forward(image, stage)
			loss = self.loss2(logit, labels) * CFG.loss2a_weight + self.loss3(logit, labels) * CFG.loss2b_weight
		
		prob2 = torch.sigmoid(logit)

		pred_mask = (prob2 > CFG.threshold).float()
		
		# print("pred_mask", pred_mask)
		
		score = fbeta_score(pred_mask, labels, threshold=CFG.threshold)

		tp, fp, fn, tn = smp.metrics.get_stats(
			pred_mask.long(), labels.long(), mode="binary"
		)
		# cur_lr = self.lr_schedulers().get_last_lr()

		return {
			"loss": loss,
			"tp": tp,
			"fp": fp,
			"fn": fn,
			"tn": tn,
			"score": score,
			# "lr": cur_lr,
		}

	def shared_epoch_end(self, outputs, stage):
		# aggregate step metics
		tp = torch.cat([x["tp"] for x in outputs])
		fp = torch.cat([x["fp"] for x in outputs])
		fn = torch.cat([x["fn"] for x in outputs])
		tn = torch.cat([x["tn"] for x in outputs])
		# lr = outputs[0]["lr"]
		loss = torch.mean(torch.Tensor([x["loss"] for x in outputs]))
		fbeta_score = torch.mean(torch.Tensor([x["score"] for x in outputs]))

		per_image_iou = smp.metrics.iou_score(
			tp, fp, fn, tn, reduction="micro-imagewise"
		)

		dataset_iou = smp.metrics.iou_score(tp, fp, fn, tn, reduction="micro")

		metrics = {
			f"{stage}_per_image_iou": per_image_iou,
			f"{stage}_dataset_iou": dataset_iou,
			f"{stage}_loss": 10000 if loss.item() == 0 else loss.item(),
			f"{stage}_tp": tp.sum().int().item(),
			f"{stage}_fp": fp.sum().int().item(),
			f"{stage}_fn": fn.sum().int().item(),
			f"{stage}_tn": tn.sum().int().item(),
			f"{stage}_score": fbeta_score.item(),
			# "lr": lr,
		}

		self.log_dict(metrics, prog_bar=True, sync_dist=True)

	def training_step(self, batch, batch_idx):
		out = self.shared_step(batch, "train")
		self.training_step_outputs.append(out)
		return out

	def on_train_epoch_end(self):
		out = self.shared_epoch_end(self.training_step_outputs, "train")
		self.training_step_outputs.clear()
		return out

	def validation_step(self, batch, batch_idx):
		out = self.shared_step(batch, "valid")
		self.validation_step_outputs.append(out)
		return out

	def on_validation_epoch_end(self):
		out = self.shared_epoch_end(self.validation_step_outputs, "valid")
		self.validation_step_outputs.clear()
		return out

	def test_step(self, batch, batch_idx):
		global predictions_map, predictions_map_counts

		patch_batch, loc_batch = batch

		loc_batch = loc_batch.long()
		patch_batch = patch_batch.float()
		predictions = self.forward(patch_batch, "test")
		predictions = predictions.sigmoid()
 
		predictions = torch.permute(predictions, (0, 2, 3, 1)).squeeze(dim=-1)
		predictions = (
			predictions.cpu().numpy()
		)
		loc_batch = loc_batch.cpu().numpy()
		
		self.test_step_outputs[0].extend(loc_batch)
		self.test_step_outputs[1].extend(predictions)        
		return loc_batch, predictions

	def on_test_epoch_end(self):
		global predictions_map, predictions_map_counts

		locs = np.array(self.test_step_outputs[0])
		preds = np.array(self.test_step_outputs[1])
		print("locs", locs.shape)
		print("preds", preds.shape)

		new_predictions_map = np.zeros_like(predictions_map[:, :, 0])
		new_predictions_map_counts = np.zeros_like(predictions_map_counts[:, :, 0])

		for (y, x), pred in zip(locs, preds):
			new_predictions_map[
				y - CFG.BUFFER : y + CFG.BUFFER, x - CFG.BUFFER : x + CFG.BUFFER
			] += pred
			new_predictions_map_counts[
				y - CFG.BUFFER : y + CFG.BUFFER, x - CFG.BUFFER : x + CFG.BUFFER
			] += 1
		
		new_predictions_map /= new_predictions_map_counts + CFG.exp
		new_predictions_map = new_predictions_map[:, :, np.newaxis]
		predictions_map = np.concatenate(
			[predictions_map, new_predictions_map], axis=-1
		)
	
	def configure_optimizers(self):
		# optimizer = optim.SGD(self.parameters(), lr=CFG.lr)
		if CFG.optimizer == "Adam":
			optimizer = optim.Adam(self.parameters(), lr=CFG.lr)
		elif CFG.optimizer == "AdamW":
			optimizer = optim.AdamW(self.parameters(), lr=CFG.lr)
		
		if CFG.lr_scheduler_name == "ReduceLROnPlateau":
			scheduler = CFG.lr_scheduler(
				optimizer, mode="min", factor=0.75, patience=5,
			)
			return {
				"optimizer": optimizer,
				"lr_scheduler": {"scheduler": scheduler, "monitor": "valid_loss"},
			}
		elif CFG.lr_scheduler_name == "CosineAnnealingLR":
			scheduler = CFG.lr_scheduler(
				optimizer, T_max=CFG.num_epochs, eta_min=CFG.eta_min_lr,
			)
			return {
				"optimizer": optimizer,
				"lr_scheduler": scheduler,
			}

In [427]:
class EnsembleModel:
    def __init__(self, test_loader, test_volume):
        super().__init__()
        self.test_loader = test_loader
        self.test_volume = test_volume
        self.list = []
        for fold in [1, 2, 3, 4]:
            _model = Model.load_from_checkpoint(
                f"weights/weights_fold-{fold}.ckpt",               
            )
            trainer = pl.Trainer(
                accelerator="gpu",
                devices="1",
                precision=16 if CFG.precision == "float16" else 32,
                enable_checkpointing=False,
            )

            self.list.append((_model, trainer))
    
    def forward(self):
        global predictions_map, predictions_map_counts
        predictions_map = np.empty_like(self.test_volume[:, :, 0])[:, :, np.newaxis].astype(np.float64)
        predictions_map_counts = np.empty_like(predictions_map).astype(np.uint16)
        for i, (model, trainer) in enumerate(self.list):
            model: Model = model
            model.test_step_outputs = [[], []]
            model.eval()
            trainer.test(
                model=model,
                dataloaders=self.test_loader,
                verbose=True,
            )
            self.list[i] = None
            gc.collect()
        predictions_map = predictions_map[:, :, 1:].mean(axis=-1, keepdims=True)

In [428]:
def opening(img):
    kernel = cv2.getStructuringElement(cv2.MORPH_ELLIPSE, (20, 20))
    return cv2.morphologyEx(img, cv2.MORPH_OPEN, kernel, iterations=2)

In [429]:


import torch
from torch.utils.data import Dataset, DataLoader
import numpy as np
from tqdm import tqdm
from skimage.transform import resize as resize_ski
import pathlib
import gc

def compute_predictions_map(split, index):
    global predictions_map
    global predictions_map_counts
    
    print(f"Load data for {split}/{index}")

    test_volume = load_volume(split=split, index=index)
    test_mask = load_mask(split=split, index=index)    
    
    #get coord
    stride = CFG.BUFFER
    H,W,D  = test_volume.shape

    ##pad #assume H,W >size
    px, py = W % stride, H % stride
    if (px != 0) or (py != 0):
        px = stride - px
        py = stride - py
        test_volume = np.pad(test_volume, [(0, py), (0, px), (0, 0)], constant_values=0)
        test_mask = np.pad(test_mask, [(0, py), (0, px)], constant_values=0)  
        
    test_locations = generate_locations_ds(test_volume, test_mask)  
    
    print(f"{len(test_locations)} test locations (before filtering by mask)")

    # filter locations inside the mask
    test_locations = [loc for loc in test_locations if is_in_masked_zone(loc, test_mask)]
    
    print(f"{len(test_locations)} test locations (after filtering by mask)")

    test_ds = SubvolumeDataset(test_locations, test_volume, None, CFG.BUFFER, is_train=False, return_location=True)
    test_loader = DataLoader(test_ds, batch_size=CFG.BATCH_SIZE, num_workers=CFG.num_workers, pin_memory=True)        

    # # shape: (Y, X, C)
    predictions_map = np.zeros_like(test_volume[:, :, 0])[:, :, np.newaxis].astype(np.float64)
    predictions_map_counts = np.zeros_like(predictions_map).astype(np.uint8)

    print("test_volume.shape", test_volume.shape)
    print("predictions_map.shape", predictions_map.shape)

    print(f"Compute predictions")
    
    model = EnsembleModel(test_loader, test_volume)
    model.forward()
    del model
    del test_locations
    del test_loader
    del test_ds
    del test_volume
    del test_mask
    gc.collect()
    
    # print("predictions_map", predictions_map, file=open("predictions_map", "w"))
    return predictions_map[:H, :W, :]


In [430]:
import numpy as np
import itertools

# (H, W)の形式
def rle(img: np.ndarray, threshold: float):
    flat_img = img.flatten()
    flat_img = np.where(flat_img > threshold, 1, 0).astype(np.uint8)

    starts = np.array((flat_img[:-1] == 0) & (flat_img[1:] == 1))
    ends = np.array((flat_img[:-1] == 1) & (flat_img[1:] == 0))

    starts_ix = np.where(starts)[0]
    ends_ix = np.where(ends)[0]

    # If the array ends in a run, we need to add a manual end index at the end of the array
    if flat_img[-1] == 1:
        ends_ix = np.append(ends_ix, flat_img.size - 1)

    starts_ix += 1  # Indices are 1-based, not 0-based
    lengths = ends_ix - starts_ix + 1

    runs = list(itertools.chain.from_iterable(zip(starts_ix, lengths)))
    
    return " ".join(map(str, runs))


In [431]:
def update_submission(predictions_map, index):
    rle_ = rle(predictions_map, threshold=CFG.threshold)
    print(f"{index}," + rle_, file=open('submission.csv', 'a'))
    # print(f"{index}," + rle_, file=open('/kaggle/working/submission.csv', 'a'))

# Submission

In [432]:
# # now = datetime.datetime.now()
# # now = f"{now}".replace(" ", "")
# now = ""
# predictions_map = None
# predictions_map_counts = None
# print("Id,Predicted", file=open('submission.csv', 'w'))
# kind = "test"
# folder = pathlib.Path(CFG.DATA_DIR) / kind
# for p in list(folder.iterdir()):
#     index = p.stem
#     predictions_map = compute_predictions_map(split=kind, index=index).squeeze(axis=-1)
#     original_size = cv2.imread(CFG.DATA_DIR + f"/{kind}/{index}/mask.png", 0).shape[:2]
#     print("original size", original_size)
#     print("predictions_map", predictions_map.shape)
#     predictions_map=xp.array(predictions_map[:, :])
#     # train/2の場合、先にリサイズするとout of memoryになる
#     predictions_map=denoise_image(predictions_map, iter_num=100)
#     predictions_map=predictions_map.get()    
# #     predictions_map = opening(predictions_map)
# #     predictions_map = resize_ski(predictions_map, (original_size[0], original_size[1]), anti_aliasing=True)
#     update_submission(predictions_map, index)
#     predictions_map = np.where(predictions_map >= CFG.threshold, 255, 0)
#     plt.imsave(f"{index}_{str(CFG.threshold)}_{str(CFG.BUFFER)}_{now}.png", predictions_map, cmap="gray")
#     del predictions_map
#     del predictions_map_counts
#     gc.collect()

# Validation

In [433]:
now = datetime.datetime.now()
now = f"{now}".replace(" ", "")
# now = ""
predictions_map = None
predictions_map_counts = None
print("Id,Predicted", file=open('submission.csv', 'w'))
kind = "train"
folder = pathlib.Path(CFG.DATA_DIR) / kind
cv_file = pathlib.Path(CFG.DATA_DIR) / kind / "1"
# for p in list(folder.iterdir()):
for p in [cv_file]:
    index = p.stem
    predictions_map = compute_predictions_map(split=kind, index=index).squeeze(axis=-1)
    original_size = cv2.imread(CFG.DATA_DIR + f"/{kind}/{index}/mask.png", 0).shape[:2]
    print("original size", original_size)
    print("predictions_map", predictions_map.shape)
    predictions_map=xp.array(predictions_map[:, :])
    # train/2の場合、先にリサイズするとout of memoryになる
    predictions_map=denoise_image(predictions_map, iter_num=100)
    predictions_map=predictions_map.get()
    # predictions_map = opening(predictions_map)
#     predictions_map = resize_ski(predictions_map, (original_size[0], original_size[1]), anti_aliasing=True)

    # Check cv results

    labels = load_labels(kind, index)
    labels = torch.from_numpy(labels)
    for thd in [0.47, 0.475, 0.495, 0.5]:
        _predictions_map = torch.from_numpy(predictions_map)
        score = fbeta_score(_predictions_map, labels, thd)
        print(f"cv score with thd={thd} of {kind}/{index}: {score}")
        __predictions_map = np.where(predictions_map >= thd, 255, 0)
        plt.imsave(f"{index}_{str(thd)}_{str(CFG.BUFFER)}_{now}.png", __predictions_map, cmap="gray")
        del _predictions_map
        del __predictions_map
        gc.collect()
    
    del predictions_map
    del predictions_map_counts
    gc.collect()

Load data for train/1


10it [00:00, 21.40it/s]


771 test locations (before filtering by mask)
771 test locations (after filtering by mask)
test_volume.shape (8256, 6336, 10)
predictions_map.shape (8256, 6336, 1)
Compute predictions


GPU available: True (cuda), used: True
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
HPU available: False, using: 0 HPUs
GPU available: True (cuda), used: True
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
HPU available: False, using: 0 HPUs
GPU available: True (cuda), used: True
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
HPU available: False, using: 0 HPUs
GPU available: True (cuda), used: True
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
HPU available: False, using: 0 HPUs
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0,1,2,3]


Testing: 0it [00:00, ?it/s]

locs (771, 2)
preds (771, 384, 384)


LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0,1,2,3]


Testing: 0it [00:00, ?it/s]

locs (771, 2)
preds (771, 384, 384)


LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0,1,2,3]


Testing: 0it [00:00, ?it/s]

locs (771, 2)
preds (771, 384, 384)


LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0,1,2,3]


Testing: 0it [00:00, ?it/s]

locs (771, 2)
preds (771, 384, 384)
original size (8181, 6330)
predictions_map (8181, 6330)


100%|██████████| 100/100 [00:39<00:00,  2.56it/s]


cv score with thd=0.47 of train/1: 0.9222993850708008
cv score with thd=0.475 of train/1: 0.9222012162208557
cv score with thd=0.495 of train/1: 0.9214496612548828
cv score with thd=0.5 of train/1: 0.9210970997810364


## Train/1

- best thd = 0.47
- best stride = 128

## Train/3

- best thd = 0.495
- best stride = 128


## Conclusion

- stride = 128 if possible

# 1-fold

## stride=192, TTA

### Train/1

- thd=0.47: 0.5684219598770142
- thd=0.475: 0.5668601989746094
- thd=0.49: 0.5621239542961121
- thd=0.495: 0.5607137680053711
- thd=0.5: 0.5592559576034546
- thd=0.51: 0.5560135245323181


### Train/3

- thd=0.475: 0.6189125180244446
- thd=0.49: 0.6188754439353943
- thd=0.495: 0.6188588738441467
- thd=0.5: 0.6188358664512634
- thd=0.51: 0.6186419725418091

## stride=128, TTA

### Train/1

- thd=0.475: 0.5764451026916504
- thd=0.49: 0.5726862549781799
- thd=0.5: 0.5697850584983826
- thd=0.51: 0.5667483806610107

### Train/3

- thd=0.475: 0.6339520215988159
- thd=0.49: 0.6342535018920898
- thd=0.495: 0.6342722773551941
- thd=0.5: 0.6343271136283875
- thd=0.51: 0.6342480778694153

# Ensemble

## stride=128, TTA

### Train/1

- thd=0.47: 0.9222993850708008
- thd=0.475: 0.9222012162208557
- thd=0.495: 0.9214496612548828
- thd=0.5: 0.9210970997810364 

### Train/3

- thd=0.475: 
- thd=0.49: 
- thd=0.495: 
- thd=0.5: 
- thd=0.51: 